### ISSUE TITLE
Race detector fails to link

### ISSUE THREAD

**Issue by tgross** [joyent/containerbuddy#75]
When I run tests with the `-race` flag to detect data races, there's what looks to be a linker failure associated with a missing glibc component. We should fix this so that we can run the race detector as part of our integration tests.

```
$ make test
docker rm -f containerbuddy_consul > /dev/null 2>&1 || true
docker run -d -m 256m --name containerbuddy_consul \
                progrium/consul:latest -server -bootstrap-expect 1 -ui-dir /ui
bf62fd69cf6a9208790d33de6536c804d6a671889cb61775e56f901d69c1b20a
docker rm -f containerbuddy_etcd > /dev/null 2>&1 || true
docker run -d -m 256m --name containerbuddy_etcd -h etcd quay.io/coreos/etcd:v2.0.8 \
                -name etcd0 \
                -advertise-client-urls http://etcd:2379,http://etcd:4001 \
                -listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 \
                -initial-advertise-peer-urls http://etcd:2380 \
                -listen-peer-urls http://0.0.0.0:2380 \
                -initial-cluster-token etcd-cluster-1 \
                -initial-cluster etcd0=http://etcd:2380 \
                -initial-cluster-state new
04fb620862a9f165f051abe0b0b9c70bbc299033e445cf2fa279b11e07201f25
docker run --rm --link containerbuddy_consul:consul --link containerbuddy_etcd:etcd \
                -v /src/containerbuddy:/go/src/containerbuddy \
                -v /src/containerbuddy/.godeps:/go/src \
                -v /src/containerbuddy/build:/build \
                -v /src/containerbuddy/cover:/cover \
                -v /src/containerbuddy/examples:/root/examples:ro \
                -v /src/containerbuddy/Makefile.docker:/go/makefile:ro \
                -e LDFLAGS='-X main.GitHash=8e9e751 -X main.Version=dev-build-not-for-release' \
                containerbuddy_build test
cd /go/src/containerbuddy && go test -v -race -coverprofile=/cover/coverage.out
# testmain
runtime/race(.text): __libc_malloc: not defined
runtime/race(.text): getuid: not defined
runtime/race(.text): pthread_self: not defined
runtime/race(.text): madvise: not defined
runtime/race(.text): madvise: not defined
runtime/race(.text): madvise: not defined
runtime/race(.text): sleep: not defined
runtime/race(.text): usleep: not defined
runtime/race(.text): abort: not defined
runtime/race(.text): isatty: not defined
runtime/race(.text): __libc_free: not defined
runtime/race(.text): getrlimit: not defined
runtime/race(.text): pipe: not defined
runtime/race(.text): __libc_stack_end: not defined
runtime/race(.text): getrlimit: not defined
runtime/race(.text): setrlimit: not defined
runtime/race(.text): setrlimit: not defined
runtime/race(.text): setrlimit: not defined
runtime/race(.text): exit: not defined
runtime/race(.text.unlikely): __errno_location: not defined
runtime/race(.text): undefined: __libc_malloc
/usr/local/go/pkg/tool/linux_amd64/link: too many errors
FAIL    containerbuddy [build failed]
makefile:36: recipe for target 'test' failed
make: *** [test] Error 2
make: *** [test] Error 2
```

**Comment by tgross** [comment]
I suspect that there are going to be some false positives in the race detection (I've seen as much before), so I'm going to propose that this can wait till post-1.0.

**Comment by tgross** [comment]
The problem we've got here is outlined in https://github.com/golang/go/issues/9918.
I can run the race detector with `CGO_ENABLED=1`, but then the build is linked against glibc and will have a dynamic library load, which we don't want to have for Containerbuddy. We could run just the race detection with `CGO_ENABLED=1` but then the results are dubious.

With those caveats, I was able to get the race detector to run. It picks up 3 races all around our SIGTERM handler -- the termination process is not atomic with all the other concurrent goroutines, and that means there's a read-on-modified-data in a couple places. But given that the signal handling is async and, more importantly, that we're seeing this data race just as we're about to flush the entire container down the toilet, I don't think there's much point in hunting this down. It's not a safety issue.

I'm going to close this issue.
